src/coprocessor/circom.rs

```diff
@@ -0,0 +1,171 @@
+// # Usage of circom coprocessors.
+//
+// See `examples/circom.rs` for a quick example of how to declare a circom coprocessor.
+
+#[cfg(not(target_arch = "wasm32"))]
+pub mod non_wasm {
+    use core::fmt::Debug;
+    use std::fs::read_dir;
+
+    use ansi_term::Colour::Red;
+    use anyhow::{bail, Result};
+    use bellperson::{ConstraintSystem, SynthesisError};
+    use circom_scotia::r1cs::CircomConfig;
+
+    use crate::{
+        circuit::gadgets::{
+            circom::CircomGadget,
+            data::GlobalAllocations,
+            pointer::{AllocatedContPtr, AllocatedPtr},
+        },
+        cli::paths::circom_dir,
+        coprocessor::{CoCircuit, Coprocessor},
+        field::LurkField,
+        ptr::Ptr,
+        store::Store,
+    };
+
+    /// To setup a new circom gadget `<NAME>`, place your circom files in a designated folder and
+    /// create a file called `<NAME>.circom`. `<CIRCOM_FOLDER>/<NAME>.circom` is the input file
+    /// for the `circom` binary; in this file you must declare your circom main component.
+    fn print_error(name: &str, available: Vec<String>) -> Result<()> {
+        let available = available.join("\n    ");
+        bail!(
+            "
+{}: no circom gadget named `{name}`.
+Available circom gadgets:
+
+    {available}
+
+If you want to setup a new circom gadget `{name}`, place your circom files in a designated folder and
+create a file called `{name}.circom`. The circom binary expects `{}_FOLDER>/{name}.circom` 
+as the input file; in this file you must declare your circom main component.
+
+Then run `lurk coprocessor --name {name} <{}_FOLDER>` to instansiate a new gadget `{name}`.",
+            Red.bold().paint("error"),
+            name.to_ascii_uppercase(),
+            name.to_ascii_uppercase(),
+        );
+    }
+
+    fn validate_gadget<F: LurkField, C: CircomGadget<F>>(gadget: &C) -> Result<()> {
+        if !circom_dir().exists() {
```

Comment:
So the workflow suggested in the example (great doc comments BTW) fails, because it reaches this point, at which the setup of the `LURK_DIRS` variable has not been achieved.

Full logs:
https://gist.github.com/huitseeker/ec7ae2675b93f2ec383a601c4376d013

Indeed, we hit the expect in `circom_dir()` because we have not called the prior required function of `set_lurk_dirs()`, as the panic message indicates. Except it's hard to call into this initialization, since at this entry point you don't have access to the cli configuration. Without a default config, the best call you can make here is:
```rust
set_lurk_dirs(&HashMap::new(), &None,& None, &None, &None);
```
We should probably have a default configuration file that would make this better - in other terms, the path configuration done in the cli's path module should have two entry points:
- one through a library, as in the calling code here, which can use env vars, a configuration file found in a conventional location, and a sane hardcoded default,
- one through the CLI, which can use the CLI's arguments on top of the above.

and most of all, *that setup should be idempotent, encapsulated, and at no point should it include any public function that can fail because function calls have not been made in the right order*.

Overall, this is not the fault of this PR, but simply because since at least #581 patently **wrong** usage of the `LURK_DIRS` `OnceCell`. Instead of asking caller to keep a state machine in their head, the path setup instead use `OnceCell::get_or_init` and not rely on error messages and testing (which in this case is non-existent, since I'm apparently the first to run through this code on a fresh state 😠 ) to hope for the programmer to use the API correctly. `OnceCell`s are meant to obviate the need for complex sequencing of initialization operations, let's use them properly.

This PR should I think dump the trivial call to `set_lurk_dirs` above. This overrides the directory setup, but I don't know how to otherwise fix this mess in a reasonable time frame -- and I will open an issue for the later cleanup.